<a href="https://colab.research.google.com/github/rewpak/AI-works/blob/main/Probabilistic_Inference_Diabetes_Na%C3%AFve_Bayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lab 4. Probabilistic Inference
# Task 4.2 Diabetes Diagnosis Using Naïve Bayes
## Problem Descriptions
Given various observations from patients, including personal details (age, gender, height, weight, etc.), body test data (blood sugar, heart rate, lipoproteins, etc.), and symptoms (fever, headache, etc.), the goal is to predict the progression level of diabetes or determine if the patient is suffering from a certain stage of the disease.

If we apply the Naïve Bayes method, we can frame the problem like this:


*   𝑃(𝑐𝑖) - prior probability of diabetes
*   𝑃(𝑐𝑖| 𝑓1,𝑓2, ... 𝑓9,𝑓10) - posterior probability (where *𝑖*  = (0-150), (150-300) (300+))
*   𝑃(𝑓1 |𝑐𝑖), 𝑃(𝑓2 |𝑐𝑖),.... 𝑃(𝑓9 |𝑐𝑖), 𝑃(𝑓10 |𝑐𝑖) conditional probability (the probability of having a disease given certain symptoms)




## Implementation and Results

In [ ]:
!pip install sklearn
import numpy as np
from sklearn import datasets
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.model_selection import train_test_split, cross_validate

# import numpy as np
import matplotlib.pyplot as plt
# from matplotlib import patches
import math

In [ ]:
diabetes = datasets.load_diabetes()
# X = diabetes.data[:,[2,3,9]]
X = diabetes.data
Y = [math.floor(x/150) for x in diabetes.target]

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.25)

nb = GaussianNB()
nb.fit(X_train, Y_train)
Y_pred = nb.predict(X_test)
acc = accuracy_score(Y_test, Y_pred)
cm = confusion_matrix(Y_test, Y_pred)
cr = classification_report(Y_test, Y_pred)

print("Accuracy:", acc)
print("Confusion Matrix:\n", cm)
print("Prior:\n", nb.class_prior_)
print("Mean:\n", nb.theta_)
variances = np.var(X_train, axis=0)
print("Variances:\n", variances)


Accuracy: 0.7027027027027027
Confusion Matrix:
 [[49 16  0]
 [12 29  2]
 [ 0  3  0]]
Prior:
 [0.52265861 0.44410876 0.03323263]
Mean:
 [[-0.00853817  0.00109077 -0.01899641 -0.01706049 -0.00727161 -0.00385798
   0.01140874 -0.01366047 -0.02110572 -0.01319276]
 [ 0.00958396  0.00269502  0.01864971  0.02226281  0.0075479   0.00415663
  -0.0139227   0.01354131  0.02467209  0.01343377]
 [ 0.0004296  -0.00131357  0.08403631  0.01655162  0.00205855  0.00392424
  -0.03804574  0.03967629  0.0382141   0.04862759]]
Variances:
 [0.00241233 0.00226988 0.00225533 0.00228095 0.00220101 0.00226726
 0.00214976 0.00216642 0.00214254 0.0022625 ]


## Discussions

In this lab, we employed a Naïve Bayes classification approach to predict the progression levels of diabetes based on various patient observations. These observations included personal details such as age, gender, height, weight, as well as body test data like blood sugar, heart rate, and lipoproteins, along with reported symptoms like fever and headache.

The model's accuracy was 70.27%, indicating decent predictive ability. It correctly predicted 49 out of 65 cases in the "0-150" category. However, it had some difficulty distinguishing between "0-150" and "150-300", misclassifying 16 cases.

In the dataset, "0-150" was the most common category (about 52%), followed by "150-300" (around 44%) while "300+" was rare (about 3%).